src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+fn font_get_glyph(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    codepoint: u32,
+    uvs: u32,
+) -> u32 {
+    let Some(codepoint) = char::from_u32(codepoint) else {
+        return 0;
+    };
+
+    if uvs == 0 {
+        caller.data().font.glyph_index(codepoint).unwrap().0 as u32
+    } else {
+        let Some(uvs) = char::from_u32(uvs) else {
+            return 0;
+        };
+        caller
+            .data()
+            .font
+            .glyph_variation_index(codepoint, uvs)
+            .unwrap_or_default()
+            .0 as u32
+    }
+}
+
+// fn font_get_scale(font: u32, x_scale: *mut i32, y_scale: *mut i32);
+// Returns the scale of the current font.
+// Just return the upem as rustybuzz has no scale.
+fn font_get_scale(mut caller: Caller<'_, ShapingData>, _font: u32, x_scale: u32, y_scale: u32) {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+    let upem = caller.data().font.units_per_em();
+
+    memory.write(
+        &mut caller.as_context_mut(),
+        x_scale as usize,
+        &upem.to_le_bytes(),
+    );
+    memory.write(
+        &mut caller.as_context_mut(),
+        y_scale as usize,
+        &upem.to_le_bytes(),
+    );
+}
+
+// fn font_get_glyph_extents(font: u32, glyph: u32, extents: *mut CGlyphExtents) -> bool;
+// Returns the glyph's extents for the given glyph ID at current scale and variation settings.
+fn font_get_glyph_extents(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    extents: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut glyph_extents = hb_glyph_extents_t::default();
+    let ret = caller
+        .data()
+        .font
+        .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+    if ret {
+        memory.write(
+            caller.as_context_mut(),
+            extents as usize,
+            bytemuck::bytes_of(&glyph_extents),
+        );
+    }
+
+    ret as u32
+}
+
+// fn font_glyph_to_string(font: u32, glyph: u32, str: *const u8, len: u32);
+// Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+fn font_glyph_to_string(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    str: u32,
+    len: u32,
+) {
+    // len is apparently the assigned heap memory. It seems I should not allocate more than that.
+    // Should not assume I am not writing over anything.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let name = caller
+        .data()
+        .font
+        .glyph_name(GlyphId(glyph as u16))
+        .map(ToOwned::to_owned)
+        .unwrap_or(format!("g{glyph:4}"));
+
+    let name = CString::new(name).unwrap();
+    let name = &name.as_bytes()[..len as usize]; // are names ever non ascii?
```

Comment:
Two things here:
- this will panic with index out of bounds if the `len` exceeds the actual length of the name. 
- if the `len` is less than `name.len() - 1`, it will not copy the null terminator

What you could do is `into_bytes` the name `String`, `truncate` it and then add the null-terminator manually. 

(Could also skip allocating at all when there's a name, but might be unnecessary optimization.)

Reply:
Done